Bring osu.Android to .NET 6

huoyaoyuan opened the pull request:
- [x] Requires https://github.com/ppy/osuTK/pull/75
- [x] Requires https://github.com/ppy/osu-framework/pull/5080
- [x] Blocked by https://github.com/dotnet/runtime/issues/67140

### Development requirement

Run `dotnet workload install` android to install the necessary SDK components (may require `sudo`).
Can be changed to `dotnet workload restore` once there's no old projects (Xamarin) remaining.
The default Android SDK version changed from 29 to 31.
Xamarin toolchain in Visual Studio should be still required. I haven't tested if the managed components can be compiled without Xamarin installed.

### Known issue

Realm currently only works in Release configuration. A hard crash will happen in Debug configuration.

### Compatibility check list

Audio, network and video all work fine together with framework side change.

Moq compatibility not tested.

Fastlane deployment not tested.

huoyaoyuan committed:
Convert osu.Android to .NET 6

huoyaoyuan committed:
Convert all Android projects and move shared properties.

huoyaoyuan committed:
Add dotnet workload command in CI

huoyaoyuan committed:
Restore workload in GitHub action

huoyaoyuan committed:
Use dotnet build to compile for android

huoyaoyuan commented:
The XA4301 warning is about picking linux assets for android from NativeLibs, since android is considered a child of linux. Should be resolved at framework side.

huoyaoyuan committed:
Fix typos in ci.yml

huoyaoyuan committed:
Remove warningaserror for Android. It's not clear how to resolve the Fody warnings now.

huoyaoyuan committed:
Move Android manifest to new default location.

huoyaoyuan committed:
Rewrite code from Xamarin.Essentials

huoyaoyuan commented:
The mono debugger issue can be workarounded by removing Xamarn.Essentials and reimplement it. There's no blocking issue I can find now.

huoyaoyuan committed:
Prefer OperatingSystem.IsAndroidVersionAtLeast over Build.VERSION.SdkInt

huoyaoyuan committed:
Handle ruleset assembly loading

huoyaoyuan committed:
Add missing EmbeddedResource to android tests

huoyaoyuan committed:
Also load ruleset assemblies for Android test

huoyaoyuan committed:
Handle sqlite load failure

huoyaoyuan committed:
Resolve Fody warnings.

huoyaoyuan committed:
Commit suggestion from code review

Co-authored-by: Susko3 <Susko3@protonmail.com>

huoyaoyuan committed:
Update AndroidBatteryInfo per code review

Susko3 commented on osu.Android/OsuGameActivity.cs:
```diff
@@ -72,11 +73,23 @@ protected override void OnCreate(Bundle savedInstanceState)
 
             RequestedOrientation = DefaultOrientation = isTablet ? ScreenOrientation.FullUser : ScreenOrientation.SensorLandscape;
+
+            // Currently (SDK 6.0.200), BundleAssemblies is not runnable for net6-android.
+            // The assembly files are not available as files either after native AOT.
+            // Manually load them so that they can be loaded by RulesetStore.loadFromAppDomain.
+            // REMEMBER to fully uninstall previous version every time when investigating this!
+            // Don't forget osu.Game.Tests.Android too.
+            Assembly.Load("osu.Game.Rulesets.Osu");
+            Assembly.Load("osu.Game.Rulesets.Taiko");
+            Assembly.Load("osu.Game.Rulesets.Catch");
+            Assembly.Load("osu.Game.Rulesets.Mania");
```
This should also be in ruleset specific test apps (probably load only the tested ruleset).

huoyaoyuan replied:
I'm not pretty sure if it's actually required. Debug build doesn't require this, as assemblies can be loaded from file. Only OsuGame and RulesetStore would impact from this, ruleset specific tests explicitly creates the ruleset.

Susko3 replied:
Looks like you're right, builds and runs just fine, in debug and release.

bdach commented:
Briefly tested this branch on my two devices, and got a pretty worrying exception stack trace on one of them:

```
2022-04-10 19:01:17 [error]: An unobserved error has occurred.
2022-04-10 19:01:17 [error]: System.AggregateException: TaskExceptionHolder_UnhandledException (net_http_ssl_connection_failed)
2022-04-10 19:01:17 [error]: ---> System.Net.Http.HttpRequestException: net_http_ssl_connection_failed
2022-04-10 19:01:17 [error]: ---> System.Security.Authentication.AuthenticationException: net_auth_SSPI
2022-04-10 19:01:17 [error]: ---> Interop+AndroidCrypto+SslException: Exception_WasThrown, Interop+AndroidCrypto+SslException
2022-04-10 19:01:17 [error]: Exception_EndOfInnerExceptionStack
2022-04-10 19:01:17 [error]: at System.Net.Security.SslStream.<ForceAuthenticationAsync>d__173`1[[System.Net.Security.AsyncReadWriteAdapter, System.Net.Security, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].MoveNext()
2022-04-10 19:01:17 [error]: at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions , HttpRequestMessage , Boolean , Stream , CancellationToken )
2022-04-10 19:01:17 [error]: Exception_EndOfInnerExceptionStack
2022-04-10 19:01:17 [error]: at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions , HttpRequestMessage , Boolean , Stream , CancellationToken )
2022-04-10 19:01:17 [error]: at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage , Boolean , CancellationToken )
2022-04-10 19:01:17 [error]: at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage , Boolean , CancellationToken )
2022-04-10 19:01:17 [error]: at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage )
2022-04-10 19:01:17 [error]: at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.<WaitWithCancellationAsync>d__1[[System.Net.Http.HttpConnection, System.Net.Http, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].MoveNext()
2022-04-10 19:01:17 [error]: at System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage , Boolean , CancellationToken )
2022-04-10 19:01:17 [error]: at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage , Boolean , Boolean , CancellationToken )
2022-04-10 19:01:17 [error]: at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage , Boolean , CancellationToken )
2022-04-10 19:01:17 [error]: at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage , Boolean , CancellationToken )
2022-04-10 19:01:17 [error]: at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage , HttpCompletionOption , CancellationTokenSource , Boolean , CancellationTokenSource , CancellationToken )
2022-04-10 19:01:17 [error]: at osu.Framework.IO.Network.WebRequest.Complete(Exception e)
2022-04-10 19:01:17 [error]: at osu.Framework.IO.Network.WebRequest.internalPerform(CancellationToken cancellationToken)
2022-04-10 19:01:17 [error]: at osu.Framework.IO.Network.WebRequest.PerformAsync(CancellationToken cancellationToken)
2022-04-10 19:01:17 [error]: at osu.Framework.Extensions.ExceptionExtensions.ExceptionExtensions.Rethrow(Exception exception)
2022-04-10 19:01:17 [error]: at osu.Framework.Extensions.ExceptionExtensions.ExceptionExtensions.RethrowAsSingular(AggregateException aggregateException)
2022-04-10 19:01:17 [error]: at osu.Framework.IO.Network.WebRequest.Perform()
2022-04-10 19:01:17 [error]: at osu.Game.Online.API.APIRequest.Perform(IAPIProvider api)
2022-04-10 19:01:17 [error]: at osu.Game.Online.API.APIAccess.Perform(APIRequest request)
2022-04-10 19:01:17 [error]: Exception_EndOfInnerExceptionStack
```

As you could guess, the above causes everything online to not work. I've checked and this does not happen on the latest release.

Device in question is a Samsung device, API 24. https://github.com/ppy/osu/issues/17499 possibly related?

huoyaoyuan commented:
Did you test if it reproduces on that device with old xamarin? If so, I think it could not block this PR as it's not a new issue.

bdach commented:
I believe it's a new issue specific to this branch. As I said the latest release works correctly on the affected device.

huoyaoyuan commented:
Changing `SocketsHttpHandler` to `HttpClientHandler` may make a difference, and the latter should be more closed to the old behavior. It can be also valuable to collect SSL config of our site and report to dotnet/runtime.

huoyaoyuan committed:
Merge branch 'master' into net6.0-android

huoyaoyuan commented:
Another SIGSEGV crash occurs when investigating the network issue.

Symptom:
- Crashes only in Debug build
- Crashes a short while after log in (into production server)
- Does not crash for news page, which is available without login
- If login was restored during startup, can crash before the main menu loaded

Finding the culprit. I've merged master and updated SDK. If you have old build you can share the behavior.

If you are prompted about "missing workload macos" after updating the SDK, try `dotnet workload uninstall` or `dotnet workload update`.

huoyaoyuan committed:
Use the new Android http handler type

Susko3 commented on osu.Android.props:
```diff
@@ -1,61 +1,24 @@
-    <Reference Include="Java.Interop" />
-  </ItemGroup>
   <ItemGroup>
     <PackageReference Include="ppy.osu.Game.Resources" Version="2022.304.0" />
```
```suggestion
```

Can be removed, works just fine without the explicit include.

Susko3 commented on osu.Android.props:
```diff
@@ -1,61 +1,24 @@
+    <RuntimeIdentifiers>android-x86;android-arm;android-arm64</RuntimeIdentifiers>
+    <AndroidPackageFormat>apk</AndroidPackageFormat>
+    <MandroidI18n>CJK;Mideast;Rare;West;Other;</MandroidI18n>
     <AndroidHttpClientHandlerType>Xamarin.Android.Net.AndroidClientHandler</AndroidHttpClientHandlerType>
```
Works on my device in debug and release when using [`SocketsHttpHandler`](https://github.com/ppy/osu-framework/blob/e6f58ba809a547842b01fc15c00b12b81dd8f09f/osu.Framework/IO/Network/WebRequest.cs#L152-L160) (the default).

https://discord.com/channels/188630481301012481/188630652340404224/966403406707195965

```suggestion
    <AndroidHttpClientHandlerType>Xamarin.Android.Net.AndroidMessageHandler</AndroidHttpClientHandlerType>
```

huoyaoyuan replied:
This should have no effect when using `SocketsHttpHandler`. Platform handler is only used by `HttpClientHandler`. Could you confirm that this changes behavior on yours?

Susko3 reviewed (commented):
Works well on both release and debug, tested login + beatmap dl + gameplay.

We may want to bump Realm to `10.11.1` to see how it behaves, as it has some changes related to `HttpClientHandler`:
https://www.mongodb.com/docs/realm-sdks/dotnet/latest/CHANGELOG.html#10111-2022-03-31

bdach commented:
I would hope that the realm bump is not really necessary, since I don't see why it would be making any web requests in our use case which should be fully offline. We shouldn't be using any of those features mentioned in the changelog ("user login, function calls, and remote mongodb calls").

huoyaoyuan commented:
I've prepared commits locally, but it's not fully working on my phone. I'll push the changes once workaround applied for all known issues.

huoyaoyuan committed:
Add workaround for trimming issue in http handler.

huoyaoyuan committed:
Remove more redundant items.

huoyaoyuan commented:
@bdach I've pushed on both repos. It's now working fine on both debug and release for me.
You can test if your old device works fine with SSL. Remember to update to latest SDK.

Things to mention: news, login, and doing something after login.

huoyaoyuan committed:
Add Moq to Android test where it's used

huoyaoyuan committed:
Merge branch 'master' into net6.0-android

huoyaoyuan commented:
@bdach When would you have time to test on your old device? If it works fine now, this PR can be unblocked.

bdach commented:
i am acutely aware that this is waiting for my testing, thanks. i have this in my backlog but it is pretty low priority in my book and new PRs to review just keep coming. will report back when able to catch up.

bdach committed:
Merge branch 'master' into net6.0-android

bdach committed:
Bump android realm reference in line with game

bdach commented:
tested this today, looks like the crashes are indeed fixed. so from that perspective this can be unblocked i suppose.

huoyaoyuan committed:
Include all the test sources for Android

huoyaoyuan committed:
Revert workaround that no longer needed for 6.0.300

This reverts commit 9b50fa346ff2868e160e2f8f9cd669593a180f1f.

huoyaoyuan committed:
Realms 10.12 can work by transitive reference with Android

huoyaoyuan committed:
Fix android project formatting

huoyaoyuan commented:
The minimum required SDK for Android should be 6.0.300 now. Every time when SDK updates, the `dotnet workload install` or `dotnet workload update` command needs to be invoked again. This may be solved by Visual Studio Maui workload released later. Just a remainder for you now.

I'm not sure whether fastlane can still build the project.
The APK path is now `osu/osu.Android/bin/Release/net6.0-android/sh.ppy.osulazer-Signed.apk`. Recording target framework in CI script can be painful for further maintenance. Should we set `AppendTargetFrameworkToOutputPath` to false in fastlane build?

peppy commented:
> Should we set `AppendTargetFrameworkToOutputPath` to false in fastlane build?

I think this should be fine.

huoyaoyuan commented:
I need help to update and test fastlane configuration because I don't have the environment (and time).
If there's nothing blocking, we should go on for this.

huoyaoyuan committed:
Merge branch 'master'

huoyaoyuan committed:
Enable NullabilityInfoContext support

huoyaoyuan commented:
It's basically working with merging latest master for both Debug and Release. I didn't test for every corner though.

huoyaoyuan committed:
Add global.json again to specify SDK version

huoyaoyuan committed:
Merge branch 'master' into net6.0-android

huoyaoyuan committed:
Update sqlite version for Android

huoyaoyuan committed:
Add back Xamarin.Essentials in Andoird.
The SIGSEGV issue seems no longer present.

huoyaoyuan committed:
Resolve error CA2201 trigger by Android test build

huoyaoyuan commented:
Weird CI issue for Xamarin.Essentials. We can change back to manual code if no workaround.

TheShadowEevee commented:
Xamarin.Essentials is missing from osu.Android/osu.Android.csproj, adding it back fixes the CI issues

huoyaoyuan commented:
Oops, I forgot to include the file when committing. Didn't take a look at the change set.

huoyaoyuan committed:
Commit forgottin Xamarin.Essentials

frenzibyte commented on osu.Android/osu.Android.csproj:
```diff
@@ -1,73 +1,22 @@
   <ItemGroup>
-    <PackageReference Include="System.Formats.Asn1">
-      <Version>5.0.0</Version>
-    </PackageReference>
+    <PackageReference Include="Xamarin.Essentials" Version="1.7.3" />
   </ItemGroup>
```
I'm not entirely sure whether `Xamarin.Essentials` works on Android or not (it doesn't on iOS), but it should probably be replaced with [Microsoft's variant](https://www.nuget.org/packages/Microsoft.Maui.Essentials.Runtime.android) for better compatibility with the framework.

huoyaoyuan replied:
> I'm not entirely sure whether `Xamarin.Essentials` works on Android or not (it doesn't on iOS)

It works on my device.



> for better compatibility with the framework

Agreed, and probably we can use that on iOS too.

huoyaoyuan replied:
The transition to Maui essentials is very straitforward. It increases required workload on develop machine (`android` -> `android-maui`).
The output apk shouldn't increase since trimming is enabled by default. I haven't confirmed though.

peppy replied:
@frenzibyte want to make the change here in line with your PR? probably best to have both on the same page..

frenzibyte committed:
Migrate osu! iOS to .NET 6 and target iOS 13.4

frenzibyte committed:
Replace `Xamarin.Essentials` with .NET MAUI Essentials

frenzibyte committed:
Work around JIT compilation error in Realm

frenzibyte committed:
Avoid configuring web proxies on iOS for now

peppy committed:
Show error message on attempting to open a URL with an unsupported protocol

peppy committed:
Update framework

frenzibyte committed:
Update usages of `SupportsJIT` in line with framework changes

frenzibyte committed:
Remove `MtouchUseLlvm` key for now

I haven't found a way to disable LLVM, and the issue may potentially be
resolved (doesn't occur for me), so we might as well just remove the key
for now.

frenzibyte committed:
Remove transitive dependencies

frenzibyte committed:
Update workflows in line with framework changes

frenzibyte committed:
Suppress MT7091

peppy committed:
Merge branch 'net6-ios' into guard-url-protocols

peppy committed:
Update framework

peppy committed:
Update `netstandard2.1` references to `net6.0`

peppy committed:
Fix the MOTHERLOAD of undetected issues that are now visible thanks to net6.0

peppy committed:
Merge branch 'master' into guard-url-protocols

peppy committed:
Fix incorrect `ConfigureAwait` call in `ImportAsUpdate`

peppy committed:
Fix assembly location lookups

peppy committed:
Fix some more inspections

frenzibyte committed:
Pin workflow to macOS 12

frenzibyte committed:
Manually use Xcode 14.1

peppy commented on global.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "sdk": {
```
Not sure we want this? @smoogipoo

huoyaoyuan replied:
Theoretically, there are environments that can compile `net6.0` but have unfixed issues in the SDK.
If we introduce `net7.0` in any project, I think it can serve as a soft requirement for SDK version.

peppy replied:
@smoogipoo can you make a call on this one? from our brief discussion last week i think you'd want to see this removed?

smoogipoo replied:
I think I would prefer this being removed, since it's one of those things that silently changes compilation in the background that isn't easily noticeable via any tooling.

peppy commented on osu.Android/osu.Android.csproj:
```diff
@@ -1,73 +1,22 @@
-    <AndroidSupportedAbis>armeabi-v7a;x86;arm64-v8a</AndroidSupportedAbis>
-    <EnableLLVM>false</EnableLLVM> <!-- This currently causes random lockups during gameplay. https://github.com/mono/mono/issues/18973 -->
+
+    <!-- This currently causes random lockups during gameplay. https://github.com/mono/mono/issues/18973 -->
```
I wonder if this is still relevant

frenzibyte replied:
Testing with the same reproduction steps in the given issue thread, I can't reproduce this at all with LLVM enabled (assuming it's enabled by default given that `clang` is used during building).

huoyaoyuan replied:
I think it's probably not, but we can test this in a separated PR.

peppy replied:
As long as it isn't forgotten.

frenzibyte committed:
Replace `Xamarin.Essentials` with .NET MAUI Essentials

frenzibyte reviewed (commented):
Other than what's mentioned above, this looks pretty sane to me and matches iOS (minus the workarounds).

peppy committed:
Merge branch 'master' into net6.0-android

peppy committed:
Merge branch 'guard-url-protocols' into net6.0-android

peppy committed:
Fix deploys to devices that don't support fast deploy

peppy committed:
Remove `rollForward` rule for now

peppy committed:
Merge branch 'master' into net6.0-android

peppy commented:
Have removed for now. Have also tested this branch deploying to device and all seems good. I'll cherry-pick the changes across to the framework update branch in their current state.
